subsys/net/lib/lwm2m/lwm2m_obj_access_control.c

```diff
@@ -0,0 +1,433 @@
+	if (lwm2m_server_short_id_to_inst(res_inst_id) >= 0) {
+		return 0;
+	}
+	/* if that res inst id does not match any ssid's, remove it */
+	int idx = -1;
+
+	for (int i = 0; i < ARRAY_SIZE(inst); i++) {
+		if (inst[i].obj && inst[i].obj_inst_id == obj_inst_id) {
+			idx = i;
+			break;
+		}
+	}
+
+	if (idx < 0) {
+		LOG_ERR("Object instance not found - %u", obj_inst_id);
+		return -ENOENT;
+	}
+
+	for (int i = 0; i < ARRAY_SIZE(inst); i++) {
+		if (res_inst[idx][ACCESS_CONTROL_ACL_ID + i].res_inst_id == res_inst_id) {
+			res_inst[idx][ACCESS_CONTROL_ACL_ID + i].res_inst_id =
+				RES_INSTANCE_NOT_CREATED;
+			break;
+		}
+	}
+	return 0;
+}
+
+static struct lwm2m_engine_obj_inst *ac_create(uint16_t obj_inst_id)
+{
+	int index, avail = -1, i = 0, j = 0;
+
+	/* Check that there is no other instance with this ID */
+	for (index = 0; index < ARRAY_SIZE(inst); index++) {
+		if (inst[index].obj && inst[index].obj_inst_id == obj_inst_id) {
+			LOG_ERR("Can not create access control instance - "
+				"already existing: %u",
+				obj_inst_id);
+			return NULL;
+		}
+
+		/* Save first available slot index */
+		if (avail < 0 && !inst[index].obj) {
+			avail = index;
+		}
+	}
+
+	if (avail < 0) {
+		LOG_ERR("Can not create access control instance - no more room: %u", obj_inst_id);
+		return NULL;
+	}
+
+	/* Set default values */
+	(void)memset(res[avail], 0, sizeof(res[avail][0]) * ARRAY_SIZE(res[avail]));
+	init_res_instance(res_inst[avail], ARRAY_SIZE(res_inst[avail]));
+
+	/* initialize instance resource data */
+	INIT_OBJ_RES_DATA(ACCESS_CONTROL_OBJECT_ID, res[avail], i, res_inst[avail], j,
+			  &m_ac_data[avail].obj_id, sizeof(m_ac_data[avail].obj_id));
+	INIT_OBJ_RES_DATA(ACCESS_CONTROL_OBJECT_INSTANCE_ID, res[avail], i, res_inst[avail], j,
+			  &m_ac_data[avail].obj_inst_id, sizeof(m_ac_data[avail].obj_inst_id));
+	INIT_OBJ_RES(ACCESS_CONTROL_ACL_ID, res[avail], i, res_inst[avail], j, MAX_SERVER_COUNT + 1,
+		     true, false, m_ac_data[avail].acl, sizeof(m_ac_data[avail].acl[0]), NULL, NULL,
+		     write_validate_cb, NULL, NULL);
+	INIT_OBJ_RES_DATA(ACCESS_CONTROL_ACCESS_CONTROL_OWNER, res[avail], i, res_inst[avail], j,
+			  &m_ac_data[avail].ac_owner, sizeof(m_ac_data[avail].ac_owner));
+
+	inst[avail].resources = res[avail];
+	inst[avail].resource_count = i;
+
+	LOG_DBG("Create access control instance: %d", obj_inst_id);
+	return &inst[avail];
+}
+
+static int ac_control_init(const struct device *dev)
+{
+	ac_obj.obj_id = LWM2M_OBJECT_ACCESS_CONTROL_ID;
+	ac_obj.version_major = ACCESS_CONTROL_VERSION_MAJOR;
+	ac_obj.version_minor = ACCESS_CONTROL_VERSION_MINOR;
+	ac_obj.is_core = true;
+	ac_obj.fields = fields;
+	ac_obj.field_count = ARRAY_SIZE(fields);
+	ac_obj.max_instance_count = ARRAY_SIZE(inst);
+	ac_obj.create_cb = ac_create;
+	lwm2m_register_obj(&ac_obj);
+#if !IS_ENABLED(CONFIG_LWM2M_RD_CLIENT_SUPPORT_BOOTSTRAP)
+	/* 0, as the default server instance when not using bootstrap is zero */
+	add_existing_objects();
```

Comment:
Isn't this redundant, i. e. already covered by the new code added to `lwm2m_register_obj()` and `engine_register_obj_inst()`.

Reply:
I don't think so. The objects/object instances that were created before the access control object was registered do not have an ac control object instance associated with them. We could change the priority of the ac obj such that it always gets created first, but the current solution is perhaps a bit more transparent.